packages/typescript-estree/tests/lib/getProjectConfigFiles.test.ts

```diff
@@ -0,0 +1,101 @@
+}));
+
+const parseSettings = {
+  filePath: './repos/repo/packages/package/file.ts',
+  tsconfigRootDir: './repos/repo',
+};
+
+describe('getProjectConfigFiles', () => {
+  it('returns the project when given as a string', () => {
+    const project = './tsconfig.eslint.json';
+
+    const actual = getProjectConfigFiles(parseSettings, project);
+
+    expect(actual).toBe(project);
+  });
+
+  it('returns the project when given as a string array', () => {
+    const project = ['./tsconfig.eslint.json'];
+
+    const actual = getProjectConfigFiles(parseSettings, project);
+
+    expect(actual).toBe(project);
+  });
+
+  it('returns the project when given as undefined', () => {
+    const project = undefined;
+
+    const actual = getProjectConfigFiles(parseSettings, project);
+
+    expect(actual).toBe(project);
+  });
+
+  describe('when caching hits', () => {
+    beforeAll(() => {
+      Date.now = (): number => 0;
+    });
+
+    it('returns a local tsconfig.json without calling existsSync a second time', () => {
+      mockExistsSync.mockReturnValue(true);
+
+      getProjectConfigFiles(parseSettings, true);
+      const actual = getProjectConfigFiles(parseSettings, true);
+
+      expect(actual).toEqual(['repos/repo/packages/package/tsconfig.json']);
+      expect(mockExistsSync).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe('when caching misses', () => {
+    beforeAll(() => {
+      // Tricks Date.now-based caching into always calling to fs.existsSync
+      let lastDateNow = 0;
+      Date.now = (): number => (lastDateNow += 1000);
+    });
+
+    it('returns a local tsconfig.json when matched', () => {
+      mockExistsSync.mockReturnValue(true);
+
+      const actual = getProjectConfigFiles(parseSettings, true);
+
+      expect(actual).toEqual(['repos/repo/packages/package/tsconfig.json']);
+    });
+
+    it('returns a parent tsconfig.json when matched', () => {
+      mockExistsSync.mockImplementation(
+        filePath => filePath === 'repos/repo/tsconfig.json',
+      );
+
+      const actual = getProjectConfigFiles(parseSettings, true);
+
+      expect(actual).toEqual(['repos/repo/tsconfig.json']);
+    });
+
+    it('throws when searching hits .', () => {
+      mockExistsSync.mockReturnValue(false);
+
+      expect(() =>
+        getProjectConfigFiles(parseSettings, true),
+      ).toThrowErrorMatchingInlineSnapshot(
+        `"project was set to \`true\` but couldn't find any tsconfig.json relative to './repos/repo/packages/package/file.ts' within './repos/repo'."`,
+      );
+    });
+
+    it('throws when searching passes the tsconfigRootDir', () => {
+      mockExistsSync.mockReturnValue(false);
+
+      expect(() =>
+        getProjectConfigFiles({ ...parseSettings, tsconfigRootDir: '/' }, true),
+      ).toThrowErrorMatchingInlineSnapshot(
+        `"project was set to \`true\` but couldn't find any tsconfig.json relative to './repos/repo/packages/package/file.ts' within '/'."`,
+      );
+    });
+  });
+});
```

Comment:
@bradzacher do you think I should add tests somewhere else too? Maybe `parse.test.ts`? I'm unsure on what best practice is here - those feel like integration / smokescreen tests?

Reply:
really that's the only way we can test this logic - with some form of physical disk structure.
you could test by mocking `fs.` but IDK doing it like that seems like you're skipping out on the most interesting part of the codebase (the actual interaction with the disk).

unit testing this logic just feels like we'll do something wrong and miss some weird filesystem behaviour and then ship bad code.

Reply:
Ok, added a few tests. I didn't want to add anything testing timing at this high a level, but it does test resolving a sibling project, resolving a parent project, and throwing an error for not finding one.